server/historian/packages/historian-base/src/routes/utils.ts

```diff
@@ -60,31 +60,64 @@ export function handleResponse<T>(
 	);
 }
 
+export function isInitial(request: Request): boolean | undefined {
+	// request.query type is { [string]: string } but it's actually { [string]: any }
+	// Account for possibilities of undefined, boolean, or string types. A number will be false.
+	return typeof request.query.initial === "undefined"
+		? undefined
+		: typeof request.query.initial === "boolean"
+			? request.query.initial
+			: request.query.initial === "true";
+}
+
+
+export class createGitServiceArgs {
+	config: nconf.Provider;
+	tenantId: string;
+	authorization: string;
+	tenantService: ITenantService;
+	storageNameProvider: IStorageNameProvider;
+	cache?: ICache;
+	asyncLocalStorage?: AsyncLocalStorage<string>;
+	initialUpload?: boolean = false;
+	allowDisabledTenant?: boolean = false;
+}
+
 export async function createGitService(
-	config: nconf.Provider,
-	tenantId: string,
-	authorization: string,
-	tenantService: ITenantService,
-	cache?: ICache,
-	asyncLocalStorage?: AsyncLocalStorage<string>,
-	allowDisabledTenant = false,
+	createArgs: createGitServiceArgs,
 ): Promise<RestGitService> {
+	const {
+		config,
+		tenantId,
+		authorization,
+		tenantService,
+		storageNameProvider,
+		cache,
+		asyncLocalStorage,
+		initialUpload,
+		allowDisabledTenant,
+	} = createArgs;
 	const token = parseToken(tenantId, authorization);
+	const decoded = decode(token) as ITokenClaims;
+	const documentId = decoded.documentId;
+	if (containsPathTraversal(documentId)) {
+		// Prevent attempted directory traversal.
+		throw new NetworkError(400, `Invalid document id: ${documentId}`);
+	}
 	const details = await tenantService.getTenant(tenantId, token, allowDisabledTenant);
 	const customData: ITenantCustomDataExternal = details.customData;
 	const writeToExternalStorage = !!customData?.externalStorageData;
-	const storageName = customData?.storageName;
-	const decoded = decode(token) as ITokenClaims;
 	const storageUrl = config.get("storageUrl") as string | undefined;
-	if (containsPathTraversal(decoded.documentId)) {
-		// Prevent attempted directory traversal.
-		throw new NetworkError(400, `Invalid document id: ${decoded.documentId}`);
-	}
+	const storageName = (
+		initialUpload
+			? await storageNameProvider?.assignStorageName(tenantId, documentId)
+			: await storageNameProvider?.retrieveStorageName(tenantId, documentId))
+		?? customData?.storageName;
```

Comment:
Above block the core logic of storage name. 

- When we do initial upload/first write, we tried to assign a storage name, which will b a load balanced injection from FRS. In OSS, here is essentially read from riddler.
- When we do further operations, we will retrieve from alfred to get document details